### PR TITLE
feat: new error ui for csv uploads

### PR DESCRIPTION
**Pull request by disKeith**
## Context

This PR implements the new error UI for bulk issue CSV uploads which includes (less password validation):
1. CSV header validation as part of the attachment upload
2. Updated error copies to match [Figma](https://www.figma.com/file/KVFsGATLUfvPqKE2lY9niH/LetterSG?type=design&node-id=732-26296&mode=design&t=xTRuK4cTl4TLqP0e-0) design

[Notion Task](https://www.notion.so/opengov/2ccadbea5c734aa09614b4ca8cc7177e?v=76cff9fca57b496f857c990e2639af2f&p=4aee5af1ff4e406e9f239a1fbda22b56&pm=s)

## Approach

1. Abstracted error messages to `addMissingParamErrorMessage` and `addExtraParamErrorMessage` for upload error messages
2. Added logic to check for extra and missing parameters in uploaded csv by comparing to the selected template's `fields`
4. Added box to house error message components in `BulkIssueUploadCsvCard`
5. Updated error message copies
6. Updated test cases

## After Screenshots
**Template Used**

> CSV used for upload will be listed in this manner

image

**Empty CSV Error**
empty_csv

**Extra Headers**
> keywords,curly,braces,extra1,Phone Number,extra2
> 1,2,3,4,5,6

extra_headers

**Password column missing (Password enabled)**
> keywords,curly,braces
> 1,2,3

missing_password

**Password column added (Password disabled)**
> keywords,curly,braces,Password
> 1,2,3,456

extra_password

**Phone Number column missing (SMS enabled)**
> keywords,curly,braces
> 1,2,3

missing_phone

**Phone Number column added (SMS disabled)**
> keywords,curly,braces,Phone Number
> 1,2,3,98765432

extra_phone

**One or more template parameter (headers) is missing**
> curly
> 1

missing_headers

**Phone number cell validation**
> keywords,curly,braces,Phone Number
> 1,2,3,91234567
> 4,5,6,66667777
> 7,8,9,98765432
> 0,1,2,9234567

cell_validate_phone_1
cell_validate_phone_2

**One or more cell content missing**
> keywords,curly,braces,Phone Number
> 1,2,3,91234567
> 4,5,6
> 7,
> 8,9,10
> 11,12,13,98765432

cell_missing_field_1
cell_missing_field_2

**One or more cell content extra**
> keywords,curly,braces,Phone Number
> 1,2,3,91234567,extra1
> 4,5,6,98765432
> 7,8,9,91112222,extra2,extra3,extra4
> 10,11,12,98765432

cell_extra_field_1
cell_extra_field_2

### CODE REVIEW

File: frontend/src/features/issue/BulkIssueDrawer/UploadCsvErrorsTable.tsx [hunk 1]
```diff
@@ -73,6 +73,8 @@ export const UploadCsvErrorsTable = ({
     )
   })
 
+  console.log(uploadCsvErrors)
```
Comment (PikkaPikkachu) [inline frontend/src/features/issue/BulkIssueDrawer/UploadCsvErrorsTable.tsx:76]: nit : we can remove this? :)

File: shared/src/util/__tests__/templates.spec.ts [hunk 2]
```diff
@@ -49,15 +49,6 @@ describe('getTemplateFields', () => {
 })
 
 describe('parseTemplateField', () => {
-  test('replaces "&nbsp;" with an empty string', () => {
```
Comment (PikkaPikkachu) [inline shared/src/util/__tests__/templates.spec.ts:52]: quick question — why do we not need this test anymore?

File: frontend/src/features/issue/components/BulkIssueUploadCsvCard.tsx [hunk 3]
```diff
@@ -135,35 +135,47 @@ export const BulkIssueUploadCsvCard = ({
...
-    if (!file || !parsedArr) return []
-
+    const addMissingParamErrorMessage = (missingParam: string) => {
+      errorMessageArr.push(`"${missingParam}" is missing in the file`)
+    }
+    const addExtraParamErrorMessage = (extraParam: string) => {
+      errorMessageArr.push(
+        `"${extraParam}" is an extra column. Please delete the column and re-upload the file`,
+      )
+    }
```
Comment (PikkaPikkachu) [inline frontend/src/features/issue/components/BulkIssueUploadCsvCard.tsx:145]: tiny nit: we can move some logic to a utils file too, if it helps keep this one easier to maintain? Up to you!

**Comment by PikkaPikkachu** [comment]
Overall LGTM! Thank you for the super comprehensive UI testing, it's very helpful!

File: frontend/src/features/issue/components/BulkIssueUploadCsvCard.tsx (hunk 3 above)
Comment (gweiying) [reply frontend/src/features/issue/components/BulkIssueUploadCsvCard.tsx:145]: Personally I think let's optimise later! For now it'll save us some navigating through files to find out what a function does

File: frontend/src/features/issue/components/BulkIssueUploadCsvCard.tsx [hunk 4]
```diff
@@ -135,35 +135,47 @@ export const BulkIssueUploadCsvCard = ({
...
+    const parsedKeys = Object.keys(parsedArr[0])
+    template.fields
+      .filter((field) => !parsedKeys.includes(field))
+      .forEach((missingParam) => addMissingParamErrorMessage(missingParam))
+    parsedKeys
+      .filter(
+        (key) => key !== '__parsed_extra' && !template.fields.includes(key),
+      )
+      .forEach((extraParam) => addExtraParamErrorMessage(extraParam))
+
```
Comment (gweiying) [inline frontend/src/features/issue/components/BulkIssueUploadCsvCard.tsx:176]: What do you think about moving this check to within `useParseCsv`?

**Review by gweiying — APPROVED** [review]
overall looks good to me! some small inline comments from @PikkaPikkachu and myself, I think mostly non-blocking

File: shared/src/util/__tests__/templates.spec.ts (hunk 2 above)
Comment (disKeith) [reply shared/src/util/__tests__/templates.spec.ts:52]: this test is no longer needed as we're no longer using `parseTemplateField` to replace nbsp entities anymore as the replacement is now handled in `TemplatesParsingService` ref: [#148](https://github.com/opengovsg/letters/pull/148)

File: frontend/src/features/issue/BulkIssueDrawer/UploadCsvErrorsTable.tsx (hunk 1 above)
Comment (disKeith) [reply frontend/src/features/issue/BulkIssueDrawer/UploadCsvErrorsTable.tsx:76]: oops

**Pull request merged**